### DISCUSSION TITLE
Can EQLP be used under Linux?

### DISCUSSION THREAD

**Discussion by jensgreven** [kauffman12/EQLogParser#221] [General]
First attempts did not look promising - used “Bottles” (running on Linux Mint 22), EQ seems to run as far as i can tell atm (servers down...), dotnetdesktop 8.0 is installed

**Comment by kauffman12** [comment]
WPF applications need a lot of workarounds and it's using some 3rd party libraries. It may not be possible.

**Comment by InkblotAdmirer** [comment]
I am currently running v2.2.24 under Linux, with Wine 9.22.  Create a 64-bit prefix, install sapi and corefonts with winetricks.  Track down espeak-1.48.04 and install in the prefix.  EDIT: (You likely don't need espeak -- selecting x86 audio won't work so no speech features. It's possible I still have this on my install notes as it may prevent a crash on launching.)

After v2.2.24 either UI or audio changes cause a fatal error when launching. I've been attempting to debug and get working, unsuccesffully as of yet.  I would like to get this working eventually in case there are breaking changes in logs, but some are still using gamparse successfully...

**Comment by kauffman12** [comment]
There were a bunch of audio changes around that time. For one, I had been using an API on the SpeechSynthesizer that did not exist on early versions of Windows 10. I replaced that with a custom implementation that depends on NAudio. I also removed all API calls I had been doing that relied on the older MME implementation. Everything should rely on WASAPI now I believe.

I guess look into WASAPI issues first. If that turns out to be the problem I could maybe make it configurable or have it try to use a different API if there's an exception. Plus there's also some calls now to check the Windows version like:
OperatingSystem.IsWindowsVersionAtLeast(10, 0, 10240)

Which also required me bringing in a couple more dotnet 8 interop libraries. Which for some reason dotnet 8.0.11 is required now. There was some security fixes that aren't backwards compatible.

**Comment by kauffman12** [comment]
So, I got it partially working with the latest. When I upload 2.2.65 you should be able to use it again for loading parses.

The APIs that are an issue at the least are SpeechSynthesizer.AllVoices and device.AudioSessionManager. So, I can't get a list of installed voices or manage the volume for the application the normal way. I could probably have a fallback version that uses espeak as a workaround and just keeps track of the volume on its own without the session manager. But not sure when I'd be able to get to that.

![wine](https://github.com/user-attachments/assets/288a524f-06f1-4b64-b76a-5aa941b47934)

**Reply by InkblotAdmirer** [reply]
Promising, EQLP now does indeed start. When loading a log file it does crash for me, however (after appearing to successfully load the log).  It looks to be related to my triggers as deleting the triggers.db file results in the ability to review the parses as you've shown.  Appreciate the attention to this niche use case.

Here is the tail end of the log, when crashing with my triggers.db:

`2025-01-01 08:34:52,588 [1] INFO EQLogParser.MainWindow - Finished Loading Log File in 2 seconds.
2025-01-01 08:34:53,780 [1] INFO EQLogParser.AudioManager - Using audio device: PulseAudio Output.
2025-01-01 08:34:53,801 [1] ERROR EQLogParser.AudioManager - Failed to access AudioSessionManager: The method or operation is not implemented.
2025-01-01 08:34:53,801 [1] WARN EQLogParser.AudioManager - Audio is NOT Initialized!
2025-01-01 08:34:53,806 [1] ERROR EQLogParser.AudioManager - System.NotImplementedException: The method or operation is not implemented.
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at WinRT.ExceptionHelpers.BorrowRestrictedErrorInfo()
   at WinRT.ExceptionHelpers.GetExceptionForHR(Int32 hr, Boolean useGlobalErrorState, Boolean& restoredExceptionFromGlobalState)
   at WinRT.ExceptionHelpers.<ThrowExceptionForHR>g__Throw|38_0(Int32 hr)
   at WinRT.ExceptionHelpers.ThrowExceptionForHR(Int32 hr)
   at ABI.Windows.Media.SpeechSynthesis.IInstalledVoicesStaticMethods.get_DefaultVoice(IObjectReference _obj)
   at Windows.Media.SpeechSynthesis.SpeechSynthesizer.get_DefaultVoice()
   at EQLogParser.AudioManager.GetVoiceInfo(String name)
   at EQLogParser.AudioManager.CreateSpeechSynthesizer(String voice)
2025-01-01 08:34:53,922 [2] ERROR EQLogParser.App - DomainUnhandledException: The method or operation is not implemented.
2025-01-01 08:34:53,923 [2] ERROR EQLogParser.App - Thread ID: 2
2025-01-01 08:34:53,923 [2] ERROR EQLogParser.App - Thread Name: 
2025-01-01 08:34:53,923 [2] ERROR EQLogParser.App -    at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at NAudio.CoreAudioApi.AudioSessionManager.UnregisterNotifications()
   at NAudio.CoreAudioApi.AudioSessionManager.Dispose()
   at NAudio.CoreAudioApi.AudioSessionManager.Finalize()`

**Reply by kauffman12** [reply]
I guess you were using triggers that just display text?

**Reply by InkblotAdmirer** [reply]
I'm using the DPS overlay, and triggers for custom overlays (spell/cooldown timers, etc). I actually have a few triggered wav files (which I forgot about) but no text to speech -- since the conversion to 64 bit (IIRC) I haven't been able to see voices to select, and I don't think I ever got TTS to work even when i could see the voices.  No text displays from triggers, though.

**Reply by kauffman12** [reply]
Ok. I'll fix the volume to work without needing the windows session manager and get the text/timer overlays working and playing sound files shouldn't be too bad. I'll try a few things to use sapi TTS or something but that may just have to wait till later.

**Comment by kauffman12** [comment]
Version 2.2.66 should be a little better. However, TTS is still not working. It looks like wine doesn't work well with the 64bit version of either of the two standard APIs I have to work with. I'll add a fallback to execute espeak-ng in the future.

Just an FYI but here's what I've been testing with.

Installed wine-10.0-rc3 and winetricks-20240105 and then ran:

winetricks corefonts
winetricks win10
winetricks videomemorysize=1024
winetricks renderer=gdi
wine windowsdesktop-runtime-8.0.11-win-x64.exe
wine EQLogParser-install-2.2.66.exe

**Reply by InkblotAdmirer** [reply]
This is working for me, with cursory testing.  I will test with this further, the updates are much appreciated -- backups for the win!

Also, I seem to have come across information that the latest Windows TTS was working at one time in Wine but have since been broken and I have seen no indication this situation is being worked.  Passing this on but caveat -- I do not recall where I came across this or whether it is actually the case.

**Reply by kauffman12** [reply]
Good to hear the other stuff seems ok so far. And yeah, I'm not sure it ever worked and espeak sounds kinda bad. I'm still looking at options. If the old 32bit speech API works then maybe I can just make a small utility app that uses it and outputs results that I can use from the parser.

**Comment by kauffman12** [comment]
For TTS. Download the latest parser, version 2.2.68 ore newer. Then go here and download this piper-tts.zip file I made:
https://drive.google.com/file/d/1G2Ecg9sfOMxifRzrKwqySGwHoVV3tHUJ/view?usp=sharing

You'll want to unzip into a directory at Program Files\EQLogParser\piper-tts. Under that directory is where you should have piper dll files and a voices directory, etc. Then start or restart the parser and see if the voices work. It includes 4 different options.

**Reply by InkblotAdmirer** [reply]
This works, and as a voice can now be selected Advanced mode can be used as well. Very nice.
